privilege/privileges/cache.go

```diff
@@ -270,3 +274,140 @@ func decodeSetToPrivilege(s types.Set) (mysql.PrivilegeType, error) {
 	}
 	return ret, nil
 }
+
+func (record *userRecord) match(user, host string) bool {
+	return record.User == user && patternMatch(record.Host, host)
+}
+
+func (record *dbRecord) match(user, host, db string) bool {
+	return record.User == user && record.DB == db &&
+		patternMatch(record.Host, host)
+}
+
+func (record *tablesPrivRecord) match(user, host, db, table string) bool {
+	return record.User == user && record.DB == db &&
+		record.TableName == table && patternMatch(record.Host, host)
+}
+
+func (record *columnsPrivRecord) match(user, host, db, table, col string) bool {
+	return record.User == user && record.DB == db &&
+		record.TableName == table && record.ColumnName == col &&
+		patternMatch(record.Host, host)
+}
+
+// patternMatch matches "%" the same way as ".*" in regular expression, for example,
+// "10.0.%" would match "10.0.1" "10.0.1.118" ...
+func patternMatch(pattern, str string) bool {
```

Comment:
Seems it is.
But it's better use another PR to extract the code to util package, and refact the code then.